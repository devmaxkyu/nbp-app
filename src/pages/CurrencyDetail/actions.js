/** 
 * @author Zemin W.
 * @description define actions to add details of currency
 * @lastModified 2019.10.4
 * */

const ADD = 'CURRENCY_ADD';

export const actions = {
    ADD
}

export const add = data =>({
    type: ADD,
    data
})