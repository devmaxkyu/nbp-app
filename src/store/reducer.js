/** 
 * @author Zemin W.
 * @description combine reducers
 * @lastModified 2019.10.4
 * */

import { combineReducers } from 'redux'

import CurrencyDetailReducer from '../pages/CurrencyDetail/reducer'
import CurrencyListReducer from '../pages/CurrencyList/reducer'
import FavouriteReducer from '../pages/Favourite/reducer'

export default combineReducers({
    CurrencyDetailReducer,
    CurrencyListReducer,
    FavouriteReducer
})
