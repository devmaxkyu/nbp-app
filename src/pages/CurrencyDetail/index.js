/** 
 * @description page component to show currency details
 * @lastModified 2019.10.4
 * @author zemin.
 * */

import React from 'react'
import {connect} from 'react-redux'
import {getCurrencyDetail,TABLE_FORMAT} from '../../services/nbpApi'
import PropTypes from 'prop-types'
import {setCurrencyDetail} from './actions'
import {addFavourite, removeFavourite} from '../Favourite/actions'
import { ENGINE_METHOD_NONE } from 'constants'
import {Link} from 'react-router-dom'
import _ from 'lodash'

class CurrencyDetail extends React.Component {
    constructor(props){
        super(props)

        this.state = {loaded:false}
    }

    componentDidMount(){

        // this.state.setState({
        //     detail:null,            
        // })

        const {             
            match:{params:{code, table}}            
        } = this.props
        console.log('CurrencyDetail.componentDidMount', code, table)
        // get details using api function
        getCurrencyDetail(table, code, this.setCurrency)        
    }

    setCurrency = (result) => {
        const {dispatch, currency} = this.props
        dispatch(setCurrencyDetail(result))
        this.setState({loaded: true})
    }

    // handler to add currency to favourite when user click button
    handleAdd = ()=> {
        const {currency, dispatch} = this.props
        dispatch(addFavourite(currency))
    }

    // handler to remove currency from favourite when user click button
    handleRemove = ()=> {
        const {currency, dispatch} = this.props
        dispatch(removeFavourite(currency))
    }

    render() {

        const {currency: detail,favourite: {list: favouriteList}} = this.props
        if(!this.state.loaded){
            return (<p>no detail data</p>)
        }
        // check if currency exists in favourite list
        const isFavourite = _.find(favouriteList, item => item.code === detail.code)
        
        return (
            <div className="App-detail">
                <div>
                    <h4>{detail.currency}</h4>
                    <p><small>{detail.code}</small></p>

                    <div className="attr">
                        <div>
                            <span className="label">Effective Date</span>
                            <span>{detail.rates[0].effectiveDate}</span>                        
                        </div>
                        <div>
                            <span className="label">Number</span>
                            <span>{detail.rates[0].no}</span>                        
                        </div>
                        <div>                        
                            <span className="label">Rate</span>
                            <span>{detail.rates[0].mid}</span>
                        </div>
                    </div>
                    
                    
                </div>
                {isFavourite?(
                    <div>
                        <p>Already in your favourite list</p>
                        <button className="App-btn" onClick={this.handleRemove}> Remove from favourite</button>
                    </div>                    
                ):(
                    <div>
                        <button className="App-btn" onClick={this.handleAdd}> Add to favourite</button>
                    </div>                    
                )}

                <div>
                    <Link to="/" >back to list</Link>
                </div>
            </div>
            
        )
    }
}


// ract router link to get request's params
CurrencyDetail.protoTypes = {
    match: PropTypes.shape({
        params: PropTypes.shape({
            code: PropTypes.string.isRequired,
            table: PropTypes.string.isRequired
        })
    })
}

export default connect(state => ({
    currency: state.currencyDetail.currency,
    favourite: state.favourite
}))(CurrencyDetail)