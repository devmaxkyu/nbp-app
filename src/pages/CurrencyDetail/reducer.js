/** 
 * @author Zemin W.
 * @description define reducers to add details of currency
 * @lastModified 2019.10.4
 * */

import { createReducer } from 'redux-create-reducer'
import { actions } from './actions'

const initialState = {}

const reducer = createReducer(initialState, {
    [actions.SET] (state, action){
        return {
            ...state,
            currency: action.data
        }
    }
})

export default reducer
