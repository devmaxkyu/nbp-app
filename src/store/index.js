/** 
 * @author Zemin W.
 * @description create store object
 * @lastModified 2019.10.4
 * */

import { createStore, applyMiddleware } from 'redux'
import { createLogger } from 'redux-logger'
import reducer from './reducer'

const store = createStore(reducer, applyMiddleware(createLogger()))

export default store
