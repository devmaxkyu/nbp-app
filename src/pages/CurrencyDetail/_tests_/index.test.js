
import React from 'react'
import CurrencyDetail from '../../CurrencyDetail'
import renderer from 'react-test-renderer'
import {Provider} from 'react-redux'
import store from '../../../store'

import { BrowserRouter as Router} from 'react-router-dom'

jest.mock('../../../services/nbpApi')

it('Currency Detail rendering', () => {
  const routeParams = {
    params: {
      code: 'USD',
      table: 'A'
    }
  }
  const component = renderer.create(
    <Provider store = {store}><Router><CurrencyDetail match={routeParams}/></Router></Provider>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
