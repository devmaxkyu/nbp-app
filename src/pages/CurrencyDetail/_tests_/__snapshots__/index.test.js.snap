// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Currency Detail rendering 1`] = `
<div
  className="App-detail"
>
  <div>
    <h4>
      afgani (Afganistan)
    </h4>
    <p>
      <small>
        AFN
      </small>
    </p>
    <div
      className="attr"
    >
      <div>
        <span
          className="label"
        >
          Effective Date
        </span>
        <span>
          2019-10-09
        </span>
      </div>
      <div>
        <span
          className="label"
        >
          Number
        </span>
        <span>
          041/B/NBP/2019
        </span>
      </div>
      <div>
        <span
          className="label"
        >
          Rate
        </span>
        <span>
          0.050396
        </span>
      </div>
    </div>
  </div>
  <div>
    <button
      className="App-btn"
      onClick={[Function]}
    >
       Add to favourite
    </button>
  </div>
  <div>
    <a
      href="/"
      onClick={[Function]}
    >
      back to list
    </a>
  </div>
</div>
`;
